 ### Loading Dataset

In [1]:
import pandas as pd
import re

In [2]:
dataSet2015 = '..\\data\\2015\\KCPD_Crime_Data_2015_20240924.csv'
df = pd.read_csv(dataSet2015)
df.head()

Report_No Reported_Date Reported Time   From_Date From Time To_Date  \
0  150015427    03/06/2015          0:02  03/06/2015      0:02     NaN   
1   70045155    09/21/2015         14:20  01/19/2015       NaN     NaN   
2   70045155    09/21/2015         14:20  01/19/2015       NaN     NaN   
3  150057463    09/08/2015          1:17  09/08/2015      1:17     NaN   
4  150033873    05/19/2015          0:21  05/19/2015      0:21     NaN   

  To Time  Offense IBRS           Description  ...  DVFlag Invl_No  \
0     NaN     2601  90Z        Misc Violation  ...       U       1   
1     NaN      401  13A  Aggravated Assault (  ...       U       1   
2     NaN     2001  90F        Family Offense  ...       U       1   
3     NaN      702  240            Auto Theft  ...       U       1   
4     NaN     1849  35A  Possession/Sale/Dist  ...       U       1   

  Involvement  Race  Sex Age Firearm Used Flag  Latitude Longitude  \
0         VIC   NaN  NaN NaN                 N   38.9767  -94.5767   
1         VIC     B    M NaN                 N    0.0000    0.0000   
2         VIC     B    M NaN                 N    0.0000    0.0000   
3         SUS     U    U NaN                 N   39.0947  -94.5516   
4         VIC   NaN  NaN NaN                 N   39.0735  -94.5461   

                                          Location 1  
0   BROADWAY KANSAS CITY64131\r\n(38.9767, -94.5767)  
1                                    99999(0.0, 0.0)  
2                                    99999(0.0, 0.0)  
3  E PROSPECT KANSAS CITY, AV 64126\r\n(39.0947, ...  
4  VICTOR KANSAS CITY, ST 64128\r\n(39.0735, -94....  

[5 rows x 26 columns]

In [3]:
df.columns

Index(['Report_No', 'Reported_Date', 'Reported Time', 'From_Date', 'From Time',
       'To_Date', 'To Time', 'Offense', 'IBRS', 'Description', 'Beat',
       'Address', 'City', 'Zip Code', 'Rep_Dist', 'Area', 'DVFlag', 'Invl_No',
       'Involvement', 'Race', 'Sex', 'Age', 'Firearm Used Flag', 'Latitude',
       'Longitude', 'Location 1'],
      dtype='object')

Columns in the dataset are as follows:

| Column Name           | Description                                                                 |
|-----------------------|-----------------------------------------------------------------------------|
| **Report_No**          | Eindeutige Identifikationsnummer des Berichts.                              |
| **Reported_Date**      | Datum, an dem das Verbrechen gemeldet wurde.                                |
| **Reported Time**      | Uhrzeit, zu der das Verbrechen gemeldet wurde.                              |
| **From_Date**          | Datum, an dem das Verbrechen stattfand oder begann.                         |
| **From Time**          | Uhrzeit, zu der das Verbrechen stattfand oder begann.                       |
| **To_Date**            | Datum, an dem das Verbrechen endete (falls bekannt).                        |
| **To Time**            | Uhrzeit, zu der das Verbrechen endete (falls bekannt).                      |
| **Offense**            | Code des gemeldeten Verbrechens.                                            |
| **IBRS**               | Klassifizierung des Verbrechens nach dem International Crime Reporting Standard. |
| **Description**        | Beschreibung des Verbrechens.                                               |
| **Beat**               | Polizeibereich oder Zone, in dem das Verbrechen gemeldet wurde.             |
| **Address**            | Adresse, an der das Verbrechen stattfand.                                   |
| **City**               | Stadt, in der das Verbrechen stattfand.                                     |
| **Zip Code**           | Postleitzahl des Verbrechensorts.                                           |
| **Rep_Dist**           | Reporting District, die Berichterstattungsregion.                           |
| **Area**               | Gebietsbezeichnung.                                                         |
| **DVFlag**             | Kennzeichnung, ob es sich um ein häusliches Verbrechen handelt (Domestic Violence Flag). |
| **Invl_No**            | Eindeutige Identifikationsnummer der beteiligten Person.                    |
| **Involvement**        | Art der Beteiligung (z. B. Täter, Opfer).                                   |
| **Race**               | Ethnische Zugehörigkeit der beteiligten Person.                             |
| **Sex**                | Geschlecht der beteiligten Person.                                          |
| **Age**                | Alter der beteiligten Person.                                               |
| **Firearm Used Flag**  | Angabe, ob eine Schusswaffe verwendet wurde.                                |
| **Latitude**           | Geografische Breite des Verbrechensortes.                                   |
| **Longitude**          | Geografische Länge des Verbrechensortes.                                    |
| **Location 1**         | Detaillierte Ortsangabe (Straße und Koordinaten) des Verbrechens.           |


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 121901 entries, 0 to 121900
Data columns (total 26 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Report_No          121901 non-null  int64  
 1   Reported_Date      121901 non-null  object 
 2   Reported Time      121901 non-null  object 
 3   From_Date          121716 non-null  object 
 4   From Time          121632 non-null  object 
 5   To_Date            46553 non-null   object 
 6   To Time            46134 non-null   object 
 7   Offense            121901 non-null  int64  
 8   IBRS               120989 non-null  object 
 9   Description        121901 non-null  object 
 10  Beat               121711 non-null  float64
 11  Address            121875 non-null  object 
 12  City               121875 non-null  object 
 13  Zip Code           121901 non-null  int64  
 14  Rep_Dist           121681 non-null  object 
 15  Area               121681 non-null  object 
 16  DV

In [5]:
missing_values = df.isnull().sum()
print(missing_values)

Report_No                0
Reported_Date            0
Reported Time            0
From_Date              185
From Time              269
To_Date              75348
To Time              75767
Offense                  0
IBRS                   912
Description              0
Beat                   190
Address                 26
City                    26
Zip Code                 0
Rep_Dist               220
Area                   220
DVFlag                   0
Invl_No                  0
Involvement              0
Race                 16746
Sex                  16746
Age                  54971
Firearm Used Flag        0
Latitude                 0
Longitude                0
Location 1               0
dtype: int64


In [6]:
dataSet2015 = '..\\data\\2015\\KCPD_Crime_Data_2015_20240924.csv'
dataSet2016 = '..\\data\\2016\\KCPD_Crime_Data_2016_20240924.csv'
dataSet2017 = '..\\data\\2017\\KCPD_Crime_Data_2017_20240924.csv'
dataSet2018 = '..\\data\\2018\\KCPD_Crime_Data_2018_20240924.csv'
dataSet2019 = '..\\data\\2019\\KCPD_Crime_Data_2019_20240924.csv'
dataSet2020 = '..\\data\\2020\\KCPD_Crime_Data_2020_20240924.csv'
dataSet2021 = '..\\data\\2021\\KCPD_Crime_Data_2021_20240924.csv'
dataSet2022 = '..\\data\\2022\\KCPD_Crime_Data_2022_20240924.csv'
dataSet2023 = '..\\data\\2023\\KCPD_Crime_Data_2023_20240924.csv'
dataSet2024 = '..\\data\\2024\\KCPD_Crime_Data_2024_20240924.csv'

df2015 = pd.read_csv(dataSet2015)
df2016 = pd.read_csv(dataSet2016)
df2017 = pd.read_csv(dataSet2017)
df2018 = pd.read_csv(dataSet2018)
df2019 = pd.read_csv(dataSet2019)
df2020 = pd.read_csv(dataSet2020)
df2021 = pd.read_csv(dataSet2021)
df2022 = pd.read_csv(dataSet2022)
df2023 = pd.read_csv(dataSet2023)
df2024 = pd.read_csv(dataSet2024)

C:\Users\leand\AppData\Local\Temp\ipykernel_20060\29561206.py:16: DtypeWarning: Columns (0,9) have mixed types. Specify dtype option on import or set low_memory=False.
  df2019 = pd.read_csv(dataSet2019)
C:\Users\leand\AppData\Local\Temp\ipykernel_20060\29561206.py:18: DtypeWarning: Columns (13) have mixed types. Specify dtype option on import or set low_memory=False.
  df2021 = pd.read_csv(dataSet2021)
C:\Users\leand\AppData\Local\Temp\ipykernel_20060\29561206.py:19: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  df2022 = pd.read_csv(dataSet2022)


In [7]:

def getdifferentColoumns():
    col2024 = df2024.columns  # Standard columns einfach mal von 2024 genommen für das letzte Jahr

    mockdatasets = {
        "2015": df2015.columns,
        "2016": df2016.columns,
        "2017": df2017.columns,
        "2018": df2018.columns,
        "2019": df2019.columns,
        "2020": df2020.columns,
        "2021": df2021.columns,
        "2022": df2022.columns,
        "2023": df2023.columns,
        "2024": df2024.columns
    }

    mismatched_columns = {}
    same_columns = []

    for year, columns in mockdatasets.items():
        if not columns.equals(col2024):
            col_diff_2024 = set(col2024) - set(columns)
            col_diff_year = set(columns) - set(col2024)
            mismatched_columns[year] = {
                "in_2024_not_in_year": col_diff_2024,
                "in_year_not_in_2024": col_diff_year
            }
        else:
            same_columns.append(year)

    if not mismatched_columns:
        print("Alle Spalten sind gleich wie 2024.")
    else:
        for year, diffs in mismatched_columns.items():
            print(f"Jahr {year} hat unterschiedliche Spalten:")
            if diffs["in_2024_not_in_year"]:
                print(f"  In 2024, aber nicht in {year}: {', '.join(diffs['in_2024_not_in_year'])}")
            if diffs["in_year_not_in_2024"]:
                print(f"  In {year}, aber nicht in 2024: {', '.join(diffs['in_year_not_in_2024'])}")

    if same_columns:
        print(f"Folgende Jahre haben die gleichen Spalten wie 2024: {', '.join(same_columns)}")


In [8]:
getdifferentColoumns()

Jahr 2015 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2015: Location, From_Time, Reported_Time, Fire Arm Used Flag, Age_Range, To_Time
  In 2015, aber nicht in 2024: Location 1, Reported Time, Firearm Used Flag, Invl_No, Latitude, Longitude, To Time, From Time
Jahr 2016 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2016: Age_Range, Location, Fire Arm Used Flag
  In 2016, aber nicht in 2024: Location 1, Firearm Used Flag, Invl_No, Latitude, Longitude
Jahr 2017 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2017: Age_Range, Fire Arm Used Flag
  In 2017, aber nicht in 2024: Firearm Used Flag, Invl_No
Jahr 2018 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2018: Age_Range, Fire Arm Used Flag
  In 2018, aber nicht in 2024: Firearm Used Flag, Invl_No
Jahr 2019 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2019: Age_Range, Fire Arm Used Flag
  In 2019, aber nicht in 2024: Firearm Used Flag
Jahr 2020 hat unterschiedliche Spalten:
  In 2024, aber ni

### Prepare the data to merge

In [9]:
df2024.rename(columns={'Fire Arm Used Flag': 'Firearm Used Flag'}, inplace=True)
df2022.rename(columns={'Fire Arm Used Flag': 'Firearm Used Flag'}, inplace=True)
df2023.rename(columns={'Fire Arm Used Flag': 'Firearm Used Flag'}, inplace=True)
getdifferentColoumns()

Jahr 2015 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2015: Location, From_Time, Reported_Time, Age_Range, To_Time
  In 2015, aber nicht in 2024: Location 1, Reported Time, Invl_No, Latitude, Longitude, To Time, From Time
Jahr 2016 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2016: Age_Range, Location
  In 2016, aber nicht in 2024: Location 1, Longitude, Latitude, Invl_No
Jahr 2017 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2017: Age_Range
  In 2017, aber nicht in 2024: Invl_No
Jahr 2018 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2018: Age_Range
  In 2018, aber nicht in 2024: Invl_No
Jahr 2019 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2019: Age_Range
Jahr 2020 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2020: Age_Range, To_Time, Reported_Time, From_Time
  In 2020, aber nicht in 2024: Reported Time, From Time, To Time
Folgende Jahre haben die gleichen Spalten wie 2024: 2021, 2022, 2023, 2024


In [10]:
print('Age', df2024["Age"])
print('Age_Range', df2024["Age_Range"])
df2021.drop('Age_Range', axis=1, inplace=True)
df2022.drop('Age_Range', axis=1, inplace=True)
df2023.drop('Age_Range', axis=1, inplace=True)
df2024.drop('Age_Range', axis=1, inplace=True)


Age 0        33.0
1         NaN
2        29.0
3        65.0
4         NaN
         ... 
80174    28.0
80175    35.0
80176    23.0
80177    20.0
80178    40.0
Name: Age, Length: 80179, dtype: float64
Age_Range 0        25-34
1          NaN
2        25-34
3        65-74
4          NaN
         ...  
80174    25-34
80175    35-44
80176    18-24
80177    18-24
80178    35-44
Name: Age_Range, Length: 80179, dtype: object


In [11]:
getdifferentColoumns()

Jahr 2015 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2015: To_Time, Location, Reported_Time, From_Time
  In 2015, aber nicht in 2024: Location 1, Reported Time, Invl_No, Latitude, Longitude, To Time, From Time
Jahr 2016 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2016: Location
  In 2016, aber nicht in 2024: Location 1, Longitude, Latitude, Invl_No
Jahr 2017 hat unterschiedliche Spalten:
  In 2017, aber nicht in 2024: Invl_No
Jahr 2018 hat unterschiedliche Spalten:
  In 2018, aber nicht in 2024: Invl_No
Jahr 2020 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2020: To_Time, Reported_Time, From_Time
  In 2020, aber nicht in 2024: Reported Time, From Time, To Time
Folgende Jahre haben die gleichen Spalten wie 2024: 2019, 2021, 2022, 2023, 2024


In [12]:
df2015.rename(columns={'Reported Time': 'Reported_Time', 'From Time': 'From_Time', 'To Time': 'To_Time'}, inplace=True)

In [13]:
df2020.rename(columns={'Reported Time': 'Reported_Time', 'From Time': 'From_Time', 'To Time': 'To_Time'}, inplace=True)
getdifferentColoumns()

Jahr 2015 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2015: Location
  In 2015, aber nicht in 2024: Location 1, Longitude, Latitude, Invl_No
Jahr 2016 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2016: Location
  In 2016, aber nicht in 2024: Location 1, Longitude, Latitude, Invl_No
Jahr 2017 hat unterschiedliche Spalten:
  In 2017, aber nicht in 2024: Invl_No
Jahr 2018 hat unterschiedliche Spalten:
  In 2018, aber nicht in 2024: Invl_No
Folgende Jahre haben die gleichen Spalten wie 2024: 2019, 2020, 2021, 2022, 2023, 2024


In [14]:
df2015.drop('Invl_No', axis=1, inplace=True)
df2016.drop('Invl_No', axis=1, inplace=True)
df2017.drop('Invl_No', axis=1, inplace=True)
df2018.drop('Invl_No', axis=1, inplace=True)

getdifferentColoumns()

Jahr 2015 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2015: Location
  In 2015, aber nicht in 2024: Location 1, Longitude, Latitude
Jahr 2016 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2016: Location
  In 2016, aber nicht in 2024: Location 1, Longitude, Latitude
Folgende Jahre haben die gleichen Spalten wie 2024: 2017, 2018, 2019, 2020, 2021, 2022, 2023, 2024


In [15]:
df2015["Location 1"]

0          BROADWAY KANSAS CITY64131\r\n(38.9767, -94.5767)
1                                           99999(0.0, 0.0)
2                                           99999(0.0, 0.0)
3         E PROSPECT KANSAS CITY, AV 64126\r\n(39.0947, ...
4         VICTOR KANSAS CITY, ST 64128\r\n(39.0735, -94....
                                ...                        
121896    6600 INDIANA AV\r\nKANSAS CITY 64132\r\n(39.00...
121897    1000 E 8 ST\r\nKANSAS CITY 64106\r\n(39.1041, ...
121898    9300 BALES DR\r\nKANSAS CITY 64132\r\n(38.9577...
121899    1800 MAIN ST\r\nKANSAS CITY 64109\r\n(39.0917,...
121900    7500 RICHMOND AV\r\nKANSAS CITY 64138\r\n(38.9...
Name: Location 1, Length: 121901, dtype: object

In [16]:
df2016["Location 1"][0]

'10TH KANSAS CITY64106\r\n(39.1019, -94.571)'

In [17]:
df2016["Latitude"][0]

39.1019

In [18]:
df2016["Longitude"][0]

-94.571

In [19]:
df2024["Location"]

0         POINT (-94.58546458 39.103614672)
1        POINT (-94.647027706 39.245607408)
2        POINT (-94.561572985 39.168812995)
3        POINT (-82.393898878 28.048895718)
4        POINT (-94.572100389 39.280825631)
                        ...                
80174    POINT (-94.668092573 39.255583732)
80175    POINT (-94.558933021 38.921884014)
80176    POINT (-94.594658033 38.991870014)
80177    POINT (-94.504258034 38.968539992)
80178    POINT (-94.581948973 39.047453012)
Name: Location, Length: 80179, dtype: object

Leandro Koordinaten: ist immer umgekeht: als Bsp: POINT (-94.58546458 39.103614672) wäre das --> (39.103615, -94.585465) in google Maps eingegeben

In [20]:
getdifferentColoumns()

Jahr 2015 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2015: Location
  In 2015, aber nicht in 2024: Location 1, Longitude, Latitude
Jahr 2016 hat unterschiedliche Spalten:
  In 2024, aber nicht in 2016: Location
  In 2016, aber nicht in 2024: Location 1, Longitude, Latitude
Folgende Jahre haben die gleichen Spalten wie 2024: 2017, 2018, 2019, 2020, 2021, 2022, 2023, 2024


In [21]:
df2015['Location'] = df2015.apply(lambda row: f"POINT ({row['Longitude']} {row['Latitude']})", axis=1)

In [22]:
df2016['Location'] = df2016.apply(lambda row: f"POINT ({row['Longitude']} {row['Latitude']})", axis=1)

In [23]:
df2015["Location"][0:3]

0    POINT (-94.5767 38.9767)
1             POINT (0.0 0.0)
2             POINT (0.0 0.0)
Name: Location, dtype: object

In [24]:
df2015[["Location", "Latitude", "Longitude", "Location 1"]][0:3]

Location  Latitude  Longitude  \
0  POINT (-94.5767 38.9767)   38.9767   -94.5767   
1           POINT (0.0 0.0)    0.0000     0.0000   
2           POINT (0.0 0.0)    0.0000     0.0000   

                                         Location 1  
0  BROADWAY KANSAS CITY64131\r\n(38.9767, -94.5767)  
1                                   99999(0.0, 0.0)  
2                                   99999(0.0, 0.0)

In [25]:
df2016["Location"][0:3]

0     POINT (-94.571 39.1019)
1    POINT (-94.5768 39.1722)
2    POINT (-94.5314 39.1042)
Name: Location, dtype: object

In [26]:
# Filter the DataFrame for rows where location is 'POINT (0.0 0.0)'
zero_point_rows_df2015 = df2015[df2015['Location'] == 'POINT (0.0 0.0)']
zero_point_rows_df2016 = df2016[df2016['Location'] == 'POINT (0.0 0.0)']

# Display the rows where location is 'POINT (0.0 0.0)'
print("Number of POINT (0.0 0.0) in 2015:", len(zero_point_rows_df2015))
print("Number of POINT (0.0 0.0) in 2016:", len(zero_point_rows_df2016))

Number of POINT (0.0 0.0) in 2015: 26
Number of POINT (0.0 0.0) in 2016: 37


> Remove Rows with no coordinates

In [27]:
df2015 = df2015[df2015['Location'] != 'POINT (0.0 0.0)']
df2016 = df2016[df2016['Location'] != 'POINT (0.0 0.0)']

In [28]:
# Filter the DataFrame for rows where location is 'POINT (0.0 0.0)'
zero_point_rows_df2015 = df2015[df2015['Location'] == 'POINT (0.0 0.0)']
zero_point_rows_df2016 = df2016[df2016['Location'] == 'POINT (0.0 0.0)']

# Display the rows where location is 'POINT (0.0 0.0)'
print("Number of POINT (0.0 0.0) in 2015:", len(zero_point_rows_df2015))
print("Number of POINT (0.0 0.0) in 2016:", len(zero_point_rows_df2016))

Number of POINT (0.0 0.0) in 2015: 0
Number of POINT (0.0 0.0) in 2016: 0


In [29]:
getdifferentColoumns()

Jahr 2015 hat unterschiedliche Spalten:
  In 2015, aber nicht in 2024: Location 1, Longitude, Latitude
Jahr 2016 hat unterschiedliche Spalten:
  In 2016, aber nicht in 2024: Location 1, Longitude, Latitude
Folgende Jahre haben die gleichen Spalten wie 2024: 2017, 2018, 2019, 2020, 2021, 2022, 2023, 2024


In [30]:
df2015.drop('Longitude', axis=1, inplace=True)
df2015.drop('Location 1', axis=1, inplace=True)
df2015.drop('Latitude', axis=1, inplace=True)

df2016.drop('Longitude', axis=1, inplace=True)
df2016.drop('Location 1', axis=1, inplace=True)
df2016.drop('Latitude', axis=1, inplace=True)

In [31]:
getdifferentColoumns()

Alle Spalten sind gleich wie 2024.
Folgende Jahre haben die gleichen Spalten wie 2024: 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023, 2024


## Merge

In [32]:
frames = [df2015,df2016, df2017, df2018, df2019, df2020, df2021, df2022, df2023, df2024]

In [33]:
merged_df = pd.concat(frames)

In [34]:
merged_df.head()

Report_No Reported_Date Reported_Time   From_Date From_Time To_Date  \
0  150015427    03/06/2015          0:02  03/06/2015      0:02     NaN   
3  150057463    09/08/2015          1:17  09/08/2015      1:17     NaN   
4  150033873    05/19/2015          0:21  05/19/2015      0:21     NaN   
5  150061779    08/31/2015         10:28  08/31/2015      8:00     NaN   
6  150087307    12/04/2015          2:57  12/04/2015      2:57     NaN   

  To_Time Offense IBRS           Description  ... Zip Code Rep_Dist Area  \
0     NaN    2601  90Z        Misc Violation  ...    64131   PJ3229  CPD   
3     NaN     702  240            Auto Theft  ...    64126   PJ7474  EPD   
4     NaN    1849  35A  Possession/Sale/Dist  ...    64128   PJ2340  EPD   
5     NaN     801  13B  Non Aggravated Assau  ...    61109   PJ1326  CPD   
6     NaN    2601  90Z        Misc Violation  ...    64108   PJ1326  CPD   

  DVFlag Involvement Race  Sex   Age Firearm Used Flag  \
0      U         VIC  NaN  NaN   NaN                 N   
3      U         SUS    U    U   NaN                 N   
4      U         VIC  NaN  NaN   NaN                 N   
5      N         SUS    B    M  29.0                 N   
6      U         VIC  NaN  NaN   NaN                 N   

                   Location  
0  POINT (-94.5767 38.9767)  
3  POINT (-94.5516 39.0947)  
4  POINT (-94.5461 39.0735)  
5  POINT (-89.0251 42.2167)  
6  POINT (-94.5645 39.0952)  

[5 rows x 23 columns]

In [35]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1093640 entries, 0 to 80178
Data columns (total 23 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   Report_No          1093640 non-null  object 
 1   Reported_Date      1093640 non-null  object 
 2   Reported_Time      1093640 non-null  object 
 3   From_Date          1092624 non-null  object 
 4   From_Time          1092199 non-null  object 
 5   To_Date            431633 non-null   object 
 6   To_Time            390011 non-null   object 
 7   Offense            1093640 non-null  object 
 8   IBRS               1029730 non-null  object 
 9   Description        1008161 non-null  object 
 10  Beat               1089226 non-null  object 
 11  Address            1093570 non-null  object 
 12  City               1093524 non-null  object 
 13  Zip Code           1046564 non-null  object 
 14  Rep_Dist           978663 non-null   object 
 15  Area               1088156 non-null  ob

In [36]:
merged_df_missing_values = merged_df.isnull().sum()
print(merged_df_missing_values)

Report_No                 0
Reported_Date             0
Reported_Time             0
From_Date              1016
From_Time              1441
To_Date              662007
To_Time              703629
Offense                   0
IBRS                  63910
Description           85479
Beat                   4414
Address                  70
City                    116
Zip Code              47076
Rep_Dist             114977
Area                   5484
DVFlag                    0
Involvement               0
Race                 148209
Sex                  141002
Age                  387197
Firearm Used Flag         0
Location              66526
dtype: int64


> Checkout rows with empty location

In [37]:
null_rows = merged_df[merged_df['Location'].isnull()]
null_rows[["Address", "City", "Zip Code", "Location"]]

Address         City Zip Code Location
0      8500  STATE LINE RD  KANSAS CITY  64114.0      NaN
1          00  NE TUDOR LN  LEES SUMMIT      NaN      NaN
2          1200  E 63RD ST  KANSAS CITY  64132.0      NaN
3       9400  MCKINLEY AVE  KANSAS CITY  64138.0      NaN
4          700  LINWOOD BL  KANSAS CITY      NaN      NaN
...                    ...          ...      ...      ...
61546      UNKNOWN UNKNOWN      UNKNOWN      NaN      NaN
62260    6900  WALROND AVE   KANSAS CIY      NaN      NaN
63605             U00  UNK          UNK      NaN      NaN
65155  UNKNOWN and UNKNOWN  KANSAS CITY      NaN      NaN
70732  UNKNOWN and UNKNOWN  KANSAS CITY      NaN      NaN

[66526 rows x 4 columns]

In [38]:
merged_df = merged_df[merged_df['Location'].notnull()]

In [39]:
merged_df_missing_values = merged_df.isnull().sum()
print(merged_df_missing_values)

Report_No                 0
Reported_Date             0
Reported_Time             0
From_Date               979
From_Time              1398
To_Date              619397
To_Time              659244
Offense                   0
IBRS                  53962
Description           68129
Beat                   3855
Address                  51
City                     96
Zip Code              42296
Rep_Dist              96273
Area                   4729
DVFlag                    0
Involvement               0
Race                 138193
Sex                  131901
Age                  368813
Firearm Used Flag         0
Location                  0
dtype: int64


> reformat some location coordinates

In [40]:
# Function to reformat coordinates using regex
def reformat_coordinates(coord):
    # Define the regex pattern to match the latitude and longitude
    pattern = r'\(([-+]?[0-9]*\.?[0-9]+),\s*([-+]?[0-9]*\.?[0-9]+)\)'
    # Search for the pattern in the coordinate string
    match = re.search(pattern, coord)
    if match:
        # Extract latitude and longitude from the match
        latitude = match.group(1)
        longitude = match.group(2)
        # Return the formatted POINT string
        return f'POINT ({longitude} {latitude})'
    return coord  # Return original if no match found

# Apply the function to the DataFrame
merged_df['Location'] = merged_df['Location'].apply(reformat_coordinates)

In [41]:
counter = 0
for i in merged_df['Location'].unique():
    if not i.startswith("POINT"):
        counter+=1
print(counter, "rows have no coordinates or they are not formatted correctly")

8613 rows have no coordinates or they are not formatted correctly


> deleting the not correctly formatted rows

In [42]:
merged_df = merged_df[merged_df['Location'].str.startswith("POINT")]

# Export

In [43]:
merged_df.to_csv("..\\data\\mergedData\\merged_df.csv")